test/boost/trie_writer_test.cc

```diff
@@ -0,0 +1,264 @@
+            _output.insert({result, std::move(s)});
+            testlog.trace("write: {}", sz.value);
+            return result;
+        }
+        size_t page_size() const {
+            return _page_size;
+        };
+        void pad_to_page_boundary() {
+            size_t pad = bytes_left_in_page();
+            _padding += pad;
+            testlog.trace("pad: {}", pad);
+            _pos = _pos + sink_offset(pad);
+        };
+        size_t bytes_left_in_page() const {
+            return round_up(_pos.value + 1, page_size()) - _pos.value;
+        };
+        sink_pos pos() const {
+            return _pos;
+        }
+    };
+    static_assert(trie_writer_sink<trie_output_stream>);
+
+    trie_output_stream out(page_size, seed);
+    const auto max_expected_alloc = branching_factor * sizeof(writer_node) + sizeof(std::max_align_t);
+    auto wr = trie_writer(out, max_chain_len, std::bit_ceil(max_expected_alloc));
+
+    std::vector<trie_payload> payloads;
+
+    std::string_view prev = {};
+    for (size_t i = 0; i < inputs.size(); ++i) {
+        size_t depth = std::ranges::mismatch(inputs[i], prev).in1 - inputs[i].begin();
+        payloads.push_back(trie_payload(1 + i%15, object_representation(42 + i)));
+        wr.add(depth, string_as_bytes(inputs[i]).subspan(depth), payloads.back());
+        prev = inputs[i];
+    }
+
+    auto root_pos = wr.finish();
+
+    if (inputs.size() == 0) {
+        SCYLLA_ASSERT(!root_pos.valid());
+        BOOST_REQUIRE_EQUAL(out.pos().value, 0);
+        return;
+    }
+
+    assert(root_pos.valid());
+
+    std::vector<std::string> outputs;
+    std::vector<trie_payload> output_payloads;
+    struct local_state {
+        sink_pos _idx;
+        int _stage;
+    };
+    std::vector<size_t> transition_lengths;
+    std::vector<std::byte> transition_stack;
+    std::vector<local_state> stack;
+    stack.push_back({root_pos, -1});
+    while (stack.size()) {
+        auto& [idx, stage] = stack.back();
+        const auto& node = out._output.at(idx);
+        if (stage < 0) {
+            transition_stack.insert(transition_stack.end(), node.transition.begin(), node.transition.end());
+            transition_lengths.push_back(node.transition.size());
+            if (node.payload._payload_bits) {
+                outputs.push_back(std::string(bytes_as_string(transition_stack).substr(1)));
+                output_payloads.push_back(node.payload);
+            }
+            stage += 1;
+        }
+        if (stage < int(node.children.size())) {
+            stage += 1;
+            stack.push_back({node.children[stage - 1], -1});
+            continue;
+        }
+        transition_stack.resize(transition_stack.size() - transition_lengths.back());
+        transition_lengths.pop_back();
+        stack.pop_back();
+    }
+    testlog.debug("Outputs: {}", outputs);
+    SCYLLA_ASSERT(std::ranges::equal(inputs, outputs));
+    SCYLLA_ASSERT(std::ranges::equal(payloads, output_payloads));
```

Comment:
Wouldn't BOOST_REQUIRE be better here?

SCYLLA_ASSERT might change its behavior one day.

Reply:
It started as BOOST_REQUIRE, then I modified it to SCYLLA_ASSERT during debugging and I forgot to change it back. I'll change it back, since that's the convention.

That said, BOOST_REQUIRE is somewhat useless when compared to the ASSERT, because it doesn't leave coredumps and doesn't stop the debugger, it just prints an error and throws an exception. To actually debug a failed BOOST_REQUIRE, you want to turn it into an assert anyway.

Reply:
I have the debugger catch the exception that BOOST_REQUIRE throws.

Reply:
It works sometimes, (and it would work for this test too, true), but many tests throw legitimate exceptions, and then `catch throw` isn't really usable.

Reply:
I have it catch the actual exception type (in those cases). I don't claim it's good, but that's what we have now.

Reply:
Fixed.